### PR TITLE
Use default config on missing configuration section

### PR DESCRIPTION
**Pull request by aufarg**
Hi, I've made a change to make default configuration to be used if `initializationOptions` doesn't have `haskell` or `languageServerHaskell` key in it. There are some issue where the server returns `haskell-lsp: configuration parse error.`. This change shouldn't break any existing configuration because it adds safety net instead of straight failing.

Also, I actually made the same change to https://github.com/haskell/haskell-ide-engine/pull/1802 first before realizing there's a new project. I'm not sure if this can be considered as bug fix, so I'll keep it open there.

### CODE REVIEW

File: hls-plugin-api/src/Ide/Plugin/Config.hs
Comment (Ailrun) [inline hls-plugin-api/src/Ide/Plugin/Config.hs:78]: I think it should not silently fall back to default configuration. For example, if a user makes a typo and define `hakell` instead, the server won't print any message about it and just hide that it uses fall back config.

Comment (aufarg) [reply hls-plugin-api/src/Ide/Plugin/Config.hs:78]: Would be it be acceptable if we log a message on the server that says something along the line `No user configuration detected. Falling back to default configuration`?

Comment (Ailrun) [reply hls-plugin-api/src/Ide/Plugin/Config.hs:78]: Yes, I believe that would give better user experience.

Comment (aufarg) [reply hls-plugin-api/src/Ide/Plugin/Config.hs:98]: I have two solutions in mind for this, none are quite satisfactory for me:
1. Have different constructor for default configuration. We log the message when we determine the config is a default one.
2. Log the config so user can check if their config matches with what the server use.

I changed the PR to solution 1 since it's the most apt solution I can think of. The problem is, we can't use record syntax, at least to my knowledge, with this approach and it looks quite unwieldy.
For solution 2, I think it's not explicit enough for the user to notice.

Let me know if there's a better approach than these two.

Comment (jneira) [reply hls-plugin-api/src/Ide/Plugin/Config.hs:98]: @aufarg Hi, sorry for the delay, and thanks for the pr.

I think the pr makes sense but it would be nice to keep the record definition.

It would be nice too to print the actual default configuration, keys and values in the log message and show an alert with the short message `No user configuration detected. Falling back to default configuration`

**Review by jneira — CHANGES_REQUESTED** [review]

File: src/Ide/Main.hs
Comment (jneira) [inline src/Ide/Main.hs:130]: What about skip the DefaultConfig constructor, keep the actual definition and only check here that there is no client config doing a pattern matching on `Maybe`.

Comment (aufarg) [reply src/Ide/Main.hs:130]: Since we make the parsing always successful, I believe `getConfig` will never return `Nothing`.

Comment (aufarg) [reply src/Ide/Main.hs:130]: I'm starting to think that to add alert message that we fallback to default config will require a change in `haskell-lsp` instead around these [lines](https://github.com/alanz/lsp/blob/36b655d9cc538f79b14c847ca383d97ffcc18162/src/Language/LSP/Server/Processing.hs#L358-L363) here. There should be a way for us to specify the default config when parsing error occurs.

**Comment by jneira** [comment]
Hi, there is a recent open issue about that (#1196) and the fix is already here so i think we have to unblock this. 
I dont fully like the change in the `Config` data type and not sure if there could be another alternative to signal the client is providing an empty config.

// cc @Hogeyama @aurfag @Ailrun

**Comment by pepeiborra** [comment]
> Hi, there is a recent open issue about that (#1196) and the fix is already here so i think we have to unblock this.
> I dont fully like the change in the `Config` data type and not sure if there could be another alternative to signal the client is providing an empty config.
> 
> // cc @Hogeyama @aufarg @Ailrun

How do we know that this works? I see no test

File: hls-plugin-api/src/Ide/Plugin/Config.hs
Comment (pepeiborra) [inline hls-plugin-api/src/Ide/Plugin/Config.hs:53]: This is not good enough. Please restore the field names or use Haddock comments to label the field

**Comment by pepeiborra** [comment]
The parser should be able to handle the empty object, since that's a valid configuration as all the configuration flags are optional. On the other hand, if a field is the wrong type, the parser should fail and HLS should error out. 

The approach taken by this PR is reasonable, but the implementation is a bit convoluted. I would handle this by making all the configuration fields optional, which is equivalent but easier to understand.

**Comment by aufarg** [comment]
> > Hi, there is a recent open issue about that (#1196) and the fix is already here so i think we have to unblock this.
> > I dont fully like the change in the `Config` data type and not sure if there could be another alternative to signal the client is providing an empty config.
> > // cc @Hogeyama @aufarg @Ailrun
> 
> How do we know that this fix works? I see no test

Sorry I haven't write the test because I wasn't sure if it should be done like this. I'll add the tests soon.

**Comment by aufarg** [comment]
> The parser should be able to handle the empty object, since that's a valid configuration as all the configuration flags are optional. On the other hand, if a field is the wrong type, the parser should fail and HLS should error out.
> 
> The approach taken by this PR is reasonable, but the implementation is a bit convoluted. I would handle this by making all the configuration fields optional, which is equivalent but easier to understand.

I actually have a question for the suggested approach here. I'm quite new to Haskell so I'm not familiar how we do things in Haskell.

So one of the comment I received above asked to make it so if user typed a wrong configuration key, we should print a message. If we make the configuration fields all optional, how do we determine if a value of `Config` not from user input? Should we check that all the fields are `Nothing`?

I also have question regarding potential code duplication if we resolve `Config` until the value is used. Is there maybe a way to resolve the value of `Config` that will be global from HLS other than in `parseConfig`?

**Comment by pepeiborra** [comment]
> > The parser should be able to handle the empty object, since that's a valid configuration as all the configuration flags are optional. On the other hand, if a field is the wrong type, the parser should fail and HLS should error out.
> > The approach taken by this PR is reasonable, but the implementation is a bit convoluted. I would handle this by making all the configuration fields optional, which is equivalent but easier to understand.
> 
> I actually have a question for the suggested approach here. I'm quite new to Haskell so I'm not familiar how we do things in Haskell.
> 
> So one of the comment I received above asked to make it so if user typed a wrong configuration key, we should print a message. If we make the configuration fields all optional, how do we determine if a value of `Config` not from user input? Should we check that all the fields are `Nothing`?
> 

If a user makes a typo in the configuration key, this will not result in an error with the current implementation. Is that what you are asking?

> I also have question regarding potential code duplication if we resolve `Config` until the value is used. Is there maybe a way to resolve the value of `Config` that will be global from HLS other than in `parseConfig`?

What do you mean by resolve here? Perhaps this blog post is helpful: https://reasonablypolymorphic.com/blog/higher-kinded-data/

**Comment by aufarg** [comment]
> If a user makes a typo in the configuration key, this will not result in an error with the current implementation. Is that what you are asking?

Yes, and it also does not log any error message. The reason I introduced a new constructor for `Config` is to differentiate between user config and default config (when we fail to detect user config).

> What do you mean by resolve here? Perhaps this blog post is helpful: https://reasonablypolymorphic.com/blog/higher-kinded-data/

Ah this seems to be what I need here. Thanks! I suppose I can also use it to check if the `Config` is all `Nothing`, it means we did not detect any configuration from the user.

**Comment by pepeiborra** [comment]
> > If a user makes a typo in the configuration key, this will not result in an error with the current implementation. Is that what you are asking?
> 
> Yes, and it also does not log any error message. The reason I introduced a new constructor for `Config` is to differentiate between user config and default config (when we fail to detect user config).

I would prefer to keep this PR focused, and just fix the issue with `{ }`. We can consider the behaviour on ` { haskell.thisIsATypo: true}` in a separate PR.

**Comment by aufarg** [comment]
> I would prefer to keep this PR focused, and just fix the issue with `{ }`. We can consider the behaviour on ` { haskell.thisIsATypo: true}` in a separate PR.

Actually if it's just fixing the issue with `{}` I have a previous commit that keeps the record definition [here](https://github.com/haskell/haskell-language-server/commit/a125697205de6dae9903cf5487fb165707c014c9). What do you think?

**Comment by pepeiborra** [comment]
Looks good to me

**Comment by aufarg** [comment]
@pepeiborra @jneira

Sorry for the delay, took me quite a while to understand lenses for implementing the test. I keep the record definition but it doesn't print any message if the key is not detected for now.

**Review by Ailrun — APPROVED** [review]
LGTM :)

**Comment by aufarg** [comment]
Hmm, not sure why my test always timed out on (8.10.3, ubuntu-latest)

**Comment by aufarg** [comment]
It seems the server does not always send log notification after config change. I updated the test code to throw dummy response so it'll not block waiting for response.

**Comment by aufarg** [comment]
@Ailrun Were you about to merge this to master? I see that you make a merge commit to merge master to this branch instead.

**Comment by Ailrun** [comment]
I'm actually waiting for @jneira's approval. I made the merge commit to check that it still works with many other recent changes.

**Review by jneira — APPROVED** [review]
lgtm too, thanks for the pr and the patience!

**Comment by Ailrun** [comment]
It looks like some formatter tests fail. Could you check it?

**Pull request merged**
